src/main/java/org/primefaces/component/datatable/DataTable.java

```diff
@@ -665,31 +668,29 @@ public String resolveStaticField(ValueExpression expression) {
         return null;
     }
 
-    public String resolveDynamicField(ValueExpression expression) {
-
-        if (expression == null) {
+    /**
+     * Get bean's property value from a value expression.
+     * Support old syntax (e.g #{car[column.property]}) instead of #{column.property}
+     * @param exprVE
+     * @return
+     */
+    public String resolveDynamicField(ValueExpression exprVE) {
+        if (exprVE == null) {
             return null;
         }
 
         FacesContext context = getFacesContext();
         ELContext elContext = context.getELContext();
+        String exprStr = exprVE.getExpressionString();
 
-        String expressionString = expression.getExpressionString();
-
-        // old syntax compatibility
-        // #{car[column.property]}
-        // new syntax is:
-        // #{column.property} or even a method call
-        if (expressionString.startsWith("#{" + getVar() + "[")) {
-            expressionString = expressionString.substring(expressionString.indexOf('[') + 1, expressionString.indexOf(']'));
-            expressionString = "#{" + expressionString + "}";
-
-            ValueExpression dynaVE = context.getApplication()
-                    .getExpressionFactory().createValueExpression(elContext, expressionString, String.class);
-            return (String) dynaVE.getValue(elContext);
+        Matcher matcher = DYNAMIC_FIELD_REGEX.matcher(exprStr );
```

Comment:
@Rapster I feel like some Javadoc is necessary here about what the REGEX is doing and examples of what it is parsing.  It will help future devs because REGEX can be hard to distinguish what it is doing without examples.

Reply:
I added javadoc on the field iteself, and rename the regex to help understand what it does